protocol/protocol.go

```diff
@@ -0,0 +1,186 @@
+package protocol
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+// NewProtocol creates a protocol
+func NewProtocol(r io.Reader, w io.Writer) *Protocol {
+	return &Protocol{
+		R: r,
+		W: w,
+	}
+}
+
+// Protocol manages the underlying wire protocol between backend and frontend.
+type Protocol struct {
```

Comment:
what about Encode/Decode?
we wrap connection with pre-defined structured messages. i think it's more similar to encoder/decoder than reader/writer. what do you think?

Reply:
yes, this is the direction. what about codec? or service?

Reply:
after discuss it, this is not encoder/decoder.
also, service is too general.

lets use `Transport` :)